fedimint-core/src/db/mod.rs

```diff
@@ -1841,28 +1861,127 @@ pub async fn apply_migrations(
         while current_db_version < target_db_version {
             if let Some(migration) = migrations.get(&current_db_version) {
                 info!(target: LOG_DB, "Migrating module {kind} from {current_db_version} to {target_db_version}");
-                migration(&mut dbtx.to_ref_nc()).await?;
+                if let Some(module_instance_id) = module_instance_id {
+                    migration(
+                        &mut global_dbtx
+                            .to_ref_with_prefix_module_id(module_instance_id)
+                            .into_nc(),
+                    )
+                    .await?;
+                } else {
+                    migration(&mut global_dbtx.to_ref_nc()).await?;
+                }
             } else {
                 panic!("Missing migration for version {current_db_version}");
             }
 
             current_db_version.increment();
-            dbtx.insert_entry(&DatabaseVersionKey, &current_db_version)
+            global_dbtx
+                .insert_entry(
+                    &DatabaseVersionKey(module_instance_id_key),
+                    &current_db_version,
+                )
                 .await;
         }
 
         current_db_version
     } else {
-        dbtx.insert_entry(&DatabaseVersionKey, &target_db_version)
-            .await;
         target_db_version
     };
 
-    dbtx.commit_tx_result().await?;
+    global_dbtx.commit_tx_result().await?;
     info!(target: LOG_DB, "{} module db version: {}", kind, db_version);
     Ok(())
 }
 
+/// Migrates the `DatabaseVersion` from inside the module's isolated database
+/// to the global namespace. Database migrations are driven from outside of the
+/// modules so the `DatabaseVersion` should also exist outside the modules.
+/// This also fixes the key prefix conflicts with 0x50.
+pub async fn migrate_database_version(
+    db: &Database,
+    target_db_version: DatabaseVersion,
+    module_instance_id: Option<ModuleInstanceId>,
+    kind: String,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+    let mut global_dbtx = db.begin_transaction().await;
+
+    // Use 0xff for fedimint-server and the `module_instance_id` for each module
+    let key_module_instance_id = if let Some(module_instance_id) = module_instance_id {
```

Comment:
Isn't this just `module_instance_id_or_global` ?

Reply:
I wrote the function then forgot to update this :)